# 2023 Workshop MISSIoN Nanopore Bioinformatics - Day 02

## Bacterial long-read _de novo_ assembly

* Bioinformatics tools overview ([slides](https://docs.google.com/presentation/d/1I3Z2ArbBWAAm3NCVLbldcaByXvVAamyWSe2T-8Q-tDo/edit?usp=sharing))
    * some command are a bit different to the examples here
    * also keep in mind that tools are regulary updated and input parameters can change (use `--help` or `-h` to see the manual for a tool!)
* Install most of them into our environment (should be already)

### Assembly (Flye)
```bash
flye --nano-raw clean_reads.fq -o flye_output -t 8 --meta --genome-size 5M

cp flye_output/assembly.fasta assembly-long.fasta
```
[Publication](https://doi.org/10.1038/s41587-019-0072-8) | [Code](https://github.com/fenderglass/Flye)


### Visualization of assembly (Bandage)
```bash
# open the GUI
Bandage &

# load graph file generated by flye:
->  flye_output/assembly_graph.gfa
# click "draw graph"
```
[Publication](http://bioinformatics.oxfordjournals.org/content/31/20/3350) | [Code](https://rrwick.github.io/Bandage/)

__Tools that have a graphical user interface can cause problems on a cluster machine__.

Alternative:
* go to https://rrwick.github.io/Bandage
* download the correct version for your Operating system, e.g.
    * download Windows version
        * or do a `wget https://github.com/rrwick/Bandage/releases/download/v0.8.1/Bandage_Windows_v0_8_1.zip`
    * unzip 
    * start `Bandage.exe` 
    * load graph file produced by flye and "Draw graph"


