# Workshop: _De novo_ assembly and mapping

## Hands-on

We will continue first with the raw FASTQ data should be located in your folder `data/230217_GI1-4_Run23-047-filtered.fastq`. Remember, that you already length-filtered the data. Use this as an input for the _de novo_ assembly. Remember to activate your Conda environment or install the necessary tools if not available.

### _De novo_ assembly (Flye)

* use your quality-checked and filtered reads for input 
* note that we're using `--nano-hq` with newer ONT chemistry
* the output folder is called `flye_output`
* we use `--meta` to activate the "expect metagenome/uneven coverage" mode which can help to recover full plasmid sequences
* we tell the tool that the expected `--genome-size` is 5 Mbp
```bash
# run the assembly, this will take a bit time
conda activate envs/workshop
flye --nano-hq 230217_GI1-4_Run23-047-filtered.fastq -o flye_output -t 10 --meta --genome-size 5M
# the final output genome assembly will be in flye_output/assembly.fasta
```

While this is running, check the original publication and the GitHub repository of the tool:

[Publication](https://doi.org/10.1038/s41587-019-0072-8) | [Code](https://github.com/fenderglass/Flye)

### Visualization of the assembly (Bandage) -> not possible on HPC 
```bash
# open the GUI
Bandage &

# load graph file generated by flye:
# ->  flye_output/assembly_graph.gfa
# click "draw graph"
```

[Publication](http://bioinformatics.oxfordjournals.org/content/31/20/3350) | [Code](https://rrwick.github.io/Bandage/)

__Tools that have a graphical user interface can cause problems on a cluster machine__.

### Mapping (minimap2)

Now, we want to map the long reads to the assembly you calculated to visualize them.

```bash
minimap2 -ax map-ont flye_output/assembly.fasta 230217_GI1-4_Run23-047-filtered.fastq > 230217_GI1-4_Run23-047-mapping.sam
```
[Publication](https://doi.org/10.1093/bioinformatics/bty191) | [Code](https://github.com/lh3/minimap2)

Inspect the resulting SAM file. Check the [SAM format specification](https://samtools.github.io/hts-specs/SAMv1.pdf).

### Visualization of the mapping (IGV)

```bash
# first, we need to convert the SAM file into a sorted BAM file to load it subsequently in IGV
samtools view -bS 230217_GI1-4_Run23-047-mapping.sam | samtools sort -@ 4 > 230217_GI1-4_Run23-047-mapping.sorted.bam  
samtools index 230217_GI1-4_Run23-047-mapping.sorted.bam

# start IGV browser and load the assembly (FASTA) and BAM file, inspect the output
igv &
```

### Alternative: Visualization of mapping (Tablet)

```bash
# open the GUI
tablet &

# load mapping file as 'primary assembly'
# ->  230217_GI1-4_Run23-047-mapping.sam

# load assembly file as 'Reference/consensus file'
# ->  flye_output/assembly.fasta
```
[Publication](http://dx.doi.org/10.1093/bib/bbs012) | [Code](https://ics.hutton.ac.uk/tablet/)

__Alternative ways to visualize such a mapping are given by (commercial software) such as Geneious or CLC Genomic Workbench.__
